### PR TITLE
refactor project layout; add embedding models 

### PR DESCRIPTION
**Pull request by bugface**
1. refactor project layout to make all models as an individual module
2. add embedding implementation for #5

### CODE REVIEW

File: MixStaticSeq/training.py [hunk 1]
```diff
@@ -165,27 +182,27 @@ def _init_new_model(self):
...
+                 'weight_decay': self.args.weight_decay},
+                {'params': [p for n, p in self.model.named_parameters() if any(nd in n for nd in no_decay)],
+                 'weight_decay': 0.0}
+            ]
+            self.optimizer = torch.optim.AdamW(optimizer_grouped_parameters, lr=self.args.learning_rate, amsgrad=True)
         elif self.args.optim == ModelOptimizers.SGD.value:
             # using momentum SGD implementation and default momentum is set to 0.9 and use nesterov
             # high variance of the parameter updates; less stable convergence; but work with batch=1
-            self.optimizer = torch.optim.SGD(self.model.parameters(), lr=self.args.learning_rate,
-                                             momentum=0.9, nesterov=True)
```
Comment (rickydangc) [inline MixStaticSeq/training.py:188]: Should momentum always equal to 0.9?

File: MixStaticSeq/seq_ehr_model.py [hunk 2]
```diff
@@ -61,18 +70,34 @@ def __init__(self, config, model_type=ModelType.M_LSTM):
...
+            raise NotImplementedError("We only support model tlsm, lstm, gru but get {}".format(model_type.value))
 
         self.non_seq_model = NonSeqModel(
             config.nonseq_input_dim, config.nonseq_hidden_dim, config.nonseq_output_dim)
 
         self.model_type = model_type
         self.dropout_rate = config.dropout_rate
-        self.merge_layer = N.Linear(config.nonseq_output_dim+config.seq_hidden_dim, config.mix_hidden_dim)
-        self.classifier = N.Linear(config.mix_hidden_dim, config.mix_output_dim)
+        self.merge_layer = nn.Linear(config.nonseq_output_dim+config.seq_hidden_dim, config.mix_hidden_dim)
```
Comment (rickydangc) [inline MixStaticSeq/seq_ehr_model.py:85]: I'm not sure if we need multiple LSTMs in prediction model using EHRs. If we need multiple LSTMs, merge layer should be like: "nn.Linear(config.nonseq_output_dim+config.seq_hidden_dim*number_of_lstm, config.mix_hidden_dim)"

File: MixStaticSeq/task.py [hunk 3]
```diff
@@ -94,6 +108,8 @@ def main(args):
     parser.add_argument("--mix_hidden_dim", default=64, type=int, help='fully connected layer size for mix model')
     parser.add_argument("--nonseq_representation_dim", default=64, type=int,
                         help='representation dim for nonseq features')
+    parser.add_argument("--log_gradients", action='store_true',
```
Comment (rickydangc) [inline MixStaticSeq/task.py:111]: no default value?

**Review by rickydangc — COMMENTED** [review]
I think I still have to test this implementation. What sample data did you use for testing?

File: MixStaticSeq/training.py (hunk 1 above)
Comment (bugface) [reply MixStaticSeq/training.py:188]: 0.9 is the best practice; in general, we do not need to tune it.

File: MixStaticSeq/task.py (hunk 3 above)
Comment (bugface) [reply MixStaticSeq/task.py:111]: action='store_true' means default is false

File: MixStaticSeq/seq_ehr_model.py (hunk 2 above)
Comment (bugface) [reply MixStaticSeq/seq_ehr_model.py:85]: That will be a special use case, I would rather leave for the person who implements the multi-LSTM model based on the use case.  This is more as a general use cases.

**Comment by bugface** [comment]
All tests are based on TLSTM original test (the simulated dataset the TLSTM used)

File: MixStaticSeq/seq_ehr_model.py [hunk 4]
```diff
@@ -98,23 +123,40 @@ def forward(self, x=None):
...
-            # y_hat = y.type(torch.long)
-            loss = F.cross_entropy(logits, y)
+            loss = F.cross_entropy(logits, y, weight=self.sampling_weight)
         else:
             raise NotImplementedError("loss mode only support bin or mul but get {}".format(self.loss_mode.value))
 
         return loss, pred_prob, torch.argmax(pred_prob, dim=-1), m_rep
+
+
+class MixEmbeddingModel(MixModel):
```
Comment (bugface) [inline MixStaticSeq/seq_ehr_model.py:146]: This MixEmbeddingModel class required extra flags for the embedding file and embedding options.  I am thinking to separate this and MixTCNModel into an individual module.

File: MixStaticSeq/seq_ehr_model.py (hunk 2 above)
Comment (rickydangc) [reply MixStaticSeq/seq_ehr_model.py:85]: OK.

**Pull request merged**